Src/VimCore/VimBuffer.fs

```diff
@@ -682,12 +688,43 @@ type internal VimBuffer
 
     /// Switch to the desired mode
     member x.SwitchMode modeKind modeArgument =
-        _modeMap.SwitchMode modeKind modeArgument
+        // Some mode switches can implicitly modify or be modified by the current one command mode
+        let overrideModeKind =
+            if VimExtensions.IsAnyInsert modeKind then
+                // switches to an insert mode end one command mode
+                _inSelectModeOneCommand <- false
+                _vimTextBuffer.InOneTimeCommand <- None
+                modeKind
+            elif VisualKind.IsAnyVisualOrSelect modeKind then
+                // Switching from an insert mode to a visual/select mode automatically initiates one command mode
+                if VimExtensions.IsAnyInsert x.Mode.ModeKind then
+                    _inSelectModeOneCommand <- false
+                    _vimTextBuffer.InOneTimeCommand <- Some x.Mode.ModeKind
+                modeKind
+            elif modeKind = ModeKind.Normal then
+                match _vimTextBuffer.InOneTimeCommand with
+                | Some oneTimeModeKind ->
+                    _inSelectModeOneCommand <- false
+                    _vimTextBuffer.InOneTimeCommand <- None
+                    oneTimeModeKind
+                | None ->
+                    modeKind
+            else
+                modeKind
+        
+        _modeMap.SwitchMode overrideModeKind modeArgument

```

Comment:
This logic feels like it's in the wrong place. The implementation of `SwitchMode` should be pretty much what it is today: switch to the mode that is provided. Logic around taking a desired mode and changing it based on the current state should be at a higher layer.

Reply:
At some point, I think I became overly concerned about missing a case without centralizing the switching override logic.  It turned out to be quite easy to move most of the logic back to VimBuffer.ProcessOneKeyInput where it originally resided, especially with the added test net.  Only one special case was needed in SelectionChangeTracker.  This seems like the best compromise because moving all of the mode switch logic to mode-specific layers would involve touching somewhere north of 200 locations that return ModeSwitch results.

How do you feel about the one-time command logic added to SwitchPreviousMode?  Removing this logic would mean duplicating it in both VimBuffer as well as VimInterpreter.RunNormal.  My testing indicates that running a "norm" command from one-command mode switches back to the saved mode (and not to command mode, which is technically the previous mode).